### PR TITLE
Reinforced plating for all your Multi-Z mapping needs

### PR DESCRIPTION
**Pull request by NamelessFairy**
## About The Pull Request

This PR adds reinforced plating and a corresponding baseturf_helper, plating that cannot be deconstructed with the RCD and requires a few steps to degrade to regular plating.

The plating is designed to serve the same purpose as R-Walls but for verticality. It shares its heat resistance with reinforced floor and hull, and in texting it can endure a single C4 blast but not X4 assuming the floor placed on it is already removed.

It is currently is unused on the existing maps due to it being poor practice to place secure locations that would justify reinforced floors on the lower Z levels, however I have spoken to people working on maps actively at the moment and they have express interest in being able to use these floors.

The plating can be constructed by using 2 sheets of plasteel on standard plating and is disassembled using wrench > welding tool > crowbar. The first stage of deconstruction causes the bolts holding the reinforcements in place to fall to the Z level below playing a sound and leaving a cleanable decal, adding a audio-visual alert that someone is about to come through your ceiling.

UPDATE: I've added a ceiling variant of the baseturf editor, this can be placed on a lower Z level where it will modify the baseturfs of the Z level above within the original area. This will make it significantly easier to ensure that you only cover tiles you want reinforced when protecting lower Z levels.

If anyone has any recommendations for sounds please tell me and I might swap them out but I think the two I've chosen work well. Additionally if anyone is able to make a better sprite for the screws or plates then that'd be a great help but I think the current ones work well enough.
## Why It's Good For The Game

Currently Multi-Z maps have a very tight restriction on where secure areas can be put, only allowing for them to be placed on the top Z level, under more secure Z levels or in exterior satellites and coated with hulls. This is due to standard plating and/or reinforced floors are very easy to get through without warning if you bring the right tools. This PR effectively adds R-Walls but for floors allowing mappers to properly protect lower Z levels from vertical infiltration methods. This also adds a visual and audible indictor to the deconstruction of reinforced floor tiles to bring them more in line with the visuals of deconstructing a wall.
## Changelog
:cl:
add: You can now reinforce plating to protect your department from the troublemakers upstairs. Station builders might find these useful to put the stations most secure locations on the lower floors.
imageadd: added sprites for reinforced plating
code: RCD proofing has been variablized and can now be applied to any floor type instead of just reinforced floors.
/:cl:

### CODE REVIEW

**Comment by Cheshify** [comment]
This will be VERY USEFUL for the map I'm currently working on, as it allows secure areas to exist on any floor other than the top one (because anyone can very easily RCD through floors!)

**Review by Twaticus — APPROVED** [review]

File: code/game/turfs/open/floor.dm [hunk 1]
```diff
@@ -28,6 +29,8 @@
 	var/floor_tile = null
 	var/list/broken_states
 	var/list/burnt_states
+	/// Determines if you can deconstruct this with a RCD
+	var/RCD_proof = FALSE
```
Comment (Mothblocks) [inline code/game/turfs/open/floor.dm:33]: `rcd_proof`

File: code/game/turfs/open/floor.dm [hunk 2]
```diff
@@ -330,11 +333,15 @@
 			new_airlock.update_appearance()
 			return TRUE
 		if(RCD_DECONSTRUCT)
-			var/old_turf_name = name
-			if(!ScrapeAway(flags = CHANGETURF_INHERIT_AIR))
+			if(RCD_proof)
+				to_chat(user, span_warning("The flooring is too thick to be regularly deconstructed!"))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor.dm:337]:
```suggestion
				balloon_alert(user, "it's too thick!")
```

File: code/game/turfs/open/floor/plating.dm [hunk 3]
```diff
@@ -82,6 +86,23 @@
...
+			balloon_alert(user, "reinforcing plating...")
+			if(do_after(user, 12 SECONDS, target = src))
+				if(sheets.get_amount() < 2)
+					return
+				sheets.use(2)
+				playsound(src, 'sound/machines/creak.ogg', 100, TRUE)
+				src.PlaceOnTop(/turf/open/floor/plating/r_plate)
+		else
+			if(!iscyborg(user))
+				to_chat(user, span_warning("This section is too damaged to support a tile! Use a welding tool to fix the damage."))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:103]:
```suggestion
				balloon_alert(user, "too damaged, use a welding tool!")
```

File: code/game/turfs/open/floor/plating.dm [hunk 4]
```diff
@@ -82,6 +86,23 @@
...
+			else
+				to_chat(user, span_warning("This section is too damaged to support a tile! Use a welding tool or a plating repair tool to fix the damage."))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:105]:
```suggestion
				balloon_alert(user, "too damaged, use a welding or plating repair tool!")
```

File: code/game/turfs/open/floor/plating.dm [hunk 5]
```diff
@@ -152,3 +174,122 @@
 
 /turf/open/floor/plating/foam/tool_act(mob/living/user, obj/item/I, tool_type)
 	return
+
+/turf/open/floor/plating/r_plate
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:178]: Rename to `reinforced`

File: code/game/turfs/open/floor/plating.dm [hunk 6]
```diff
@@ -152,3 +174,122 @@
...
+
+	thermal_conductivity = 0.025
+	heat_capacity = INFINITY
+
+	baseturfs = /turf/open/floor/plating
+	allow_replacement = FALSE
+	RCD_proof = TRUE
+	upgradable = FALSE
+
+	var/d_state = PLATE_INTACT
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:191]: https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#use-descriptive-and-obvious-names

File: code/game/turfs/open/floor/plating.dm [hunk 7]
```diff
@@ -152,3 +174,122 @@
...
+		to_chat(user, span_warning("You don't have the dexterity to do this!"))
+		return
+
+	//get the user's location
+	if(!isturf(user.loc))
+		return //can't do this stuff whilst inside objects and such
+
+	add_fingerprint(user)
+
+	var/turf/T = user.loc
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:222]: https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#use-descriptive-and-obvious-names

File: code/game/turfs/open/floor/plating.dm [hunk 8]
```diff
@@ -152,3 +174,122 @@
...
+	if(try_decon(W, user, T))
+		return
+	return ..()
+
+/turf/open/floor/plating/r_plate/proc/try_decon(obj/item/W, mob/user, turf/T)
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:227]: https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#use-descriptive-and-obvious-names

For W, T, and for the proc name--`try_deconstruct`

File: code/game/turfs/open/floor/plating.dm [hunk 9]
```diff
@@ -152,3 +174,122 @@
...
+	//DECONSTRUCTION
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:228]: This comment is useless

File: code/game/turfs/open/floor/plating.dm [hunk 10]
```diff
@@ -152,3 +174,122 @@
...
+	switch(d_state)
+		if(PLATE_INTACT)
+			if(W.tool_behaviour == TOOL_WRENCH)
+				to_chat(user, span_notice("You start loosening the bolts which secure the reinforcing plate in place..."))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:232]:
```suggestion
				balloon_alert(user, "loosening bolts...")
```

File: code/game/turfs/open/floor/plating.dm [hunk 11]
```diff
@@ -152,3 +174,122 @@
...
+				if(W.use_tool(src, user, 100, volume=100))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:233]:
```suggestion
				if(W.use_tool(src, user, 10 SECONDS, volume=100))
```

File: code/game/turfs/open/floor/plating.dm [hunk 12]
```diff
@@ -152,3 +174,122 @@
...
+					if(!istype(src, /turf/open/floor/plating/r_plate) || d_state != PLATE_INTACT)
+						return TRUE
+					d_state = PLATE_BOLTS_LOSENED
+					update_appearance()
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:237]: If you only need to update the icon, use `update_appearance(UPDATE_ICON)`

File: code/game/turfs/open/floor/plating.dm [hunk 13]
```diff
@@ -152,3 +174,122 @@
...
+					drop_screws()
+					to_chat(user, span_notice("You remove the bolts securing the reinforced plating."))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:239]:
```suggestion
					balloon_alert(user, "removed bolts")
```

File: code/game/turfs/open/floor/plating.dm [hunk 14]
```diff
@@ -152,3 +174,122 @@
...
+				return TRUE
+
+		if(PLATE_BOLTS_LOSENED)
+			if(W.tool_behaviour == TOOL_WELDER)
+				if(!W.tool_start_check(user, amount=0))
+					return
+				to_chat(user, span_notice("You begin slicing through the reinforced plating..."))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:246]:
```suggestion
				balloon_alert(user, "slicing...")
```

File: code/game/turfs/open/floor/plating.dm [hunk 15]
```diff
@@ -152,3 +174,122 @@
...
+				if(W.use_tool(src, user, 150, volume=100))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:247]:
```suggestion
				if(W.use_tool(src, user, 15 SECONDS, volume=100))
```

File: code/game/turfs/open/floor/plating.dm [hunk 16]
```diff
@@ -152,3 +174,122 @@
...
+					if(!istype(src, /turf/open/floor/plating/r_plate) || d_state != PLATE_BOLTS_LOSENED)
+						return TRUE
+					d_state = PLATE_CUT
+					update_appearance()
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:251]: Same

File: code/game/turfs/open/floor/plating.dm [hunk 17]
```diff
@@ -152,3 +174,122 @@
...
+					to_chat(user, span_notice("You press firmly on the plating, dislodging it."))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:252]:
```suggestion
					balloon_alert(user, "sliced through")
```

File: code/game/turfs/open/floor/plating.dm [hunk 18]
```diff
@@ -152,3 +174,122 @@
...
+				return TRUE
+
+			if(W.tool_behaviour == TOOL_SCREWDRIVER)
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:255]: Should be a switch/case

File: code/game/turfs/open/floor/plating.dm [hunk 19]
```diff
@@ -152,3 +174,122 @@
...
+				to_chat(user, span_notice("You begin securing the bolts that hold the reinforced plate in place..."))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:256]:
```suggestion
				balloon_alert(user, "securing bolts...")
```

File: code/game/turfs/open/floor/plating.dm [hunk 20]
```diff
@@ -152,3 +174,122 @@
...
+				if(W.use_tool(src, user, 150, volume=100))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:257]:
```suggestion
				if(W.use_tool(src, user, 15 SECONDS, volume=100))
```

File: code/game/turfs/open/floor/plating.dm [hunk 21]
```diff
@@ -152,3 +174,122 @@
...
+					if(!istype(src, /turf/open/floor/plating/r_plate) || d_state != PLATE_BOLTS_LOSENED)
+						return TRUE
+					d_state = PLATE_INTACT
+					update_appearance()
+					to_chat(user, span_notice("The bolts have been tightly secured."))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:262]:
```suggestion
					balloon_alert(user, "secured")
```

File: code/game/turfs/open/floor/plating.dm [hunk 22]
```diff
@@ -152,3 +174,122 @@
...
+				return TRUE
+
+		if(PLATE_CUT)
+			if(W.tool_behaviour == TOOL_CROWBAR)
+				to_chat(user, span_notice("You struggle to pry off the reinforced plating..."))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:267]:
```suggestion
				balloon_alert(user, "prying off...")
```

File: code/game/turfs/open/floor/plating.dm [hunk 23]
```diff
@@ -152,3 +174,122 @@
...
+				if(W.use_tool(src, user, 200, volume=100))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:268]:
```suggestion
				if(W.use_tool(src, user, 20 SECONDS, volume=100))
```

File: code/game/turfs/open/floor/plating.dm [hunk 24]
```diff
@@ -152,3 +174,122 @@
...
+					if(!istype(src,  /turf/open/floor/plating/r_plate) || d_state != PLATE_CUT)
+						return TRUE
+					to_chat(user, span_notice("You pry off the plating reinforcements."))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:271]:
```suggestion
					balloon_alert(user, "pried off")
```

File: code/game/turfs/open/floor/plating.dm [hunk 25]
```diff
@@ -152,3 +174,122 @@
...
+					new /obj/item/stack/sheet/plasteel(src, 2)
+					ScrapeAway(flags = CHANGETURF_INHERIT_AIR)
+				return TRUE
+
+			if(W.tool_behaviour == TOOL_WELDER)
+				if(!W.tool_start_check(user, amount=0))
+					return
+				to_chat(user, span_notice("You begin welding the reinforcements back onto the plating..."))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:279]:
```suggestion
				balloon_alert(user, "welding back on...")
```

File: code/game/turfs/open/floor/plating.dm [hunk 26]
```diff
@@ -152,3 +174,122 @@
...
+				if(W.use_tool(src, user, 150, volume=100))
+					if(!istype(src,  /turf/open/floor/plating/r_plate) || d_state != PLATE_CUT)
+						return TRUE
+					d_state = PLATE_BOLTS_LOSENED
+					update_appearance()
+					to_chat(user, span_notice("The reinfoced cover has been welded securely to the plating."))
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:285]:
```suggestion
					balloon_alert(user, "welded back on")
```

File: code/game/turfs/open/floor/plating.dm [hunk 27]
```diff
@@ -152,3 +174,122 @@
...
+				return TRUE
+	return FALSE
+
+/turf/open/floor/plating/r_plate/proc/drop_screws() //When you start dismantling R-Plates they'll drop their bolts on the Z-level below, a little visible warning.
+	var/turf/below_turf = get_step_multiz(src, DOWN)
+	while(istype(below_turf, /turf/open/openspace))
+		below_turf = get_step_multiz(below_turf, DOWN)
+	if(!isnull(below_turf) && !isspaceturf(below_turf))
+		new /obj/effect/decal/cleanable/glass/plastitanium/screws(below_turf)
+		playsound(src, 'sound/effects/structure_stress/pop3.ogg', 100, TRUE)
```
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:295]:
```suggestion
		playsound(src, 'sound/effects/structure_stress/pop3.ogg', 100, vary = TRUE)
```

File: code/game/turfs/open/floor/plating.dm (hunk 8 above)
Comment (Mothblocks) [inline code/game/turfs/open/floor/plating.dm:227]: What is T even used for here?

File: code/__DEFINES/construction.dm [hunk 28]
```diff
@@ -32,6 +32,11 @@
 #define RWINDOW_BOLTS_HEATED 7
 #define RWINDOW_SECURE 8
 
+//reinforced plating deconstruction states
+#define PLATE_INTACT 0
```
Comment (Mothblocks) [inline code/__DEFINES/construction.dm:36]: If these are only used in the reinforced plating file, only define them in there

File: code/modules/mapping/mapping_helpers.dm [hunk 29]
```diff
@@ -87,6 +87,11 @@
 	name = "lavaland baseturf editor"
 	baseturf = /turf/open/lava/smooth/lava_land_surface
 
+/obj/effect/baseturf_helper/r_plate
```
Comment (Mothblocks) [inline code/modules/mapping/mapping_helpers.dm:90]: Full name

**Review by Mothblocks — CHANGES_REQUESTED** [review]
reinf_floor -> reinforced_floor, reinforced, etc. No abbreviations

**Comment by Wallemations** [comment]
I thought we already had reinforced floor? Isn't that the stuff in the engine room?

Also you got no screenshots of the tiles

**Comment by RikuTheKiller** [comment]
Yay, more stuff you can't make as engiborg.

I gotta stop slacking and roll out the engiborg stuff already, heh.

**Comment by Mothblocks** [comment]
We have reinforced floors but not reinforced plating, I think

**Comment by NamelessFairy** [comment]
> Yay, more stuff you can't make as an engiborg.
> 
> I gotta stop slacking and roll out the engiborg stuff already, heh.

These plates are meant to be the counterpart to R-Walls which afaik are also not borg constructible however I can't see any critical issues with giving borgs the ability to make these in the future.

**Comment by NamelessFairy** [comment]
> I thought we already had reinforced floor? Isn't that the stuff in the engine room?
> 
> Also you got no screenshots of the tiles

all deconstruction states accompanied by the original tile

image

**Comment by NamelessFairy** [comment]
> We have reinforced floors but not reinforced plating, I think

We have reinforced floors and they share a lot of properties with the plates like the higher resistance to heat, explosives and RCD proofing. However reinforced floors only take 3 seconds to deconstruct and their aesthetic is not suitable to be used everywhere, e.g. using R-Floors in the middle of medbay to protect a downstairs vault would look out of place. R plates take longer to deconstruct and are more expensive in both materials and time to build while also maintaining the aesthetic of upstairs floors.

**Comment by NamelessFairy** [comment]
wait why does re-requesting a review remove all the other reviewers

**Comment by Wallemations** [comment]
> > I thought we already had reinforced floor? Isn't that the stuff in the engine room?
> 
> > 
> 
> > Also you got no screenshots of the tiles
> 
> 
> 
> all deconstruction states accompanied by the original non-reinforced plate
> 
> 
> 
> image
> 
> 

So what do they look like built? How are they differentiated from other plating visually?

File: code/modules/mapping/mapping_helpers.dm [hunk 30]
```diff
@@ -87,6 +87,22 @@
...
+/obj/effect/baseturf_helper/reinforced_plating/ceiling
+	name = "reinforced ceiling plating baseturf editor"
+
+/obj/effect/baseturf_helper/reinforced_plating/ceiling/replace_baseturf(turf/thing)
+	var/turf/ceiling = get_step_multiz(thing, UP)
+	if(isnull(ceiling))
+		CRASH("baseturf helper is attempting to modify the Z level above but there is no Z level above above it.")
+	if(isspaceturf(ceiling) || istype(ceiling, /turf/open/openspace))
+		return
+	. = ..(ceiling)
```
Comment (Mothblocks) [inline code/modules/mapping/mapping_helpers.dm:104]:
```suggestion
	return ..(ceiling)
```

**Comment by Mothblocks** [comment]
File still needs to be renamed too

**Comment by NamelessFairy** [comment]
> File still needs to be renamed too

Can I just double check what your referring to with this?

**Comment by Redrover1760** [comment]
Can this be applied the nukie shuttle? If its not parked on the upper Z levels, one can dig through the ceiling without much hardship, completely avoiding the turrets.

**Comment by NamelessFairy** [comment]
> Could you use these on the nukie shuttle? If its not parked on the upper Z levels, one can dig through the ceiling without much hardship, completely avoiding the turrets, or something like that.

I want to say in theory its possible? I can't find the code where the shuttle ceiling is set up though and given its a shuttle reinforced hull might make more sense. I'd consider it beyond the scope of this PR though as I don't want to touch on existing content balance, rather just open new options for how mappers balance their maps.

**Review by Mothblocks — APPROVED** [review]

File: code/game/turfs/open/floor/plating.dm [hunk 31]
```diff
@@ -72,16 +76,33 @@
...
+			var/obj/item/stack/sheet/sheets = C
+			if(sheets.get_amount() < 2)
+				return
+			balloon_alert(user, "reinforcing plating...")
+			if(do_after(user, 12 SECONDS, target = src))
+				if(sheets.get_amount() < 2)
+					return
+				sheets.use(2)
+				playsound(src, 'sound/machines/creak.ogg', 100, vary = TRUE)
+				src.PlaceOnTop(/turf/open/floor/plating/reinforced)
```
Comment (Time-Green) [inline code/game/turfs/open/floor/plating.dm:100]: You don't need to call src here unless you think it helps for clarity

File: code/game/turfs/open/floor/plating.dm [hunk 32]
```diff
@@ -152,3 +174,133 @@
 
 /turf/open/floor/plating/foam/tool_act(mob/living/user, obj/item/I, tool_type)
 	return
+
+//reinforced plating deconstruction states
+#define PLATE_INTACT 0
+#define PLATE_BOLTS_LOSENED 1
```
Comment (Time-Green) [inline code/game/turfs/open/floor/plating.dm:180]: Typo?

File: code/game/turfs/open/floor/plating.dm [hunk 33]
```diff
@@ -152,3 +174,133 @@
...
+
+	thermal_conductivity = 0.025
+	heat_capacity = INFINITY
+
+	baseturfs = /turf/open/floor/plating
+	allow_replacement = FALSE
+	rcd_proof = TRUE
+	upgradable = FALSE
+
+	var/deconstruction_state = PLATE_INTACT
```
Comment (Time-Green) [inline code/game/turfs/open/floor/plating.dm:196]: autodoc

File: code/modules/mapping/mapping_helpers.dm [hunk 34]
```diff
@@ -87,6 +87,22 @@
 	name = "lavaland baseturf editor"
 	baseturf = /turf/open/lava/smooth/lava_land_surface
 
+/obj/effect/baseturf_helper/reinforced_plating
```
Comment (Time-Green) [inline code/modules/mapping/mapping_helpers.dm:90]: Could you autodoc these new classes?

File: code/game/objects/effects/decals/cleanable/misc.dm [hunk 35]
```diff
@@ -54,6 +54,10 @@
 /obj/effect/decal/cleanable/glass/plastitanium
 	icon_state = "plastitaniumtiny"
 
+/obj/effect/decal/cleanable/glass/plastitanium/screws //I don't know how to sprite scattered screws, this can work until a spriter gets their hands on it.
```
Comment (Time-Green) [inline code/game/objects/effects/decals/cleanable/misc.dm:57]: Could you autodoc this class?

**Review by Time-Green — CHANGES_REQUESTED** [review]
Looks good! Just nitpicks tbh. 

Try and autodoc your variables and procs though. Internal procs don't _have_ to be autodoc, but I'd still appreciate it

Helpful link: 
https://github.com/tgstation/tgstation/blob/master/.github/guides/AUTODOC.md

**Pull request merged**